group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +573,81 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
+        if (offsetsByTopic == null) {
+            return Optional.of(groupId);
+        }
+        try {
+            Group group = groupMetadataManager.group(groupId);
+            ExpirationCondition expirationCondition = group.expirationCondition();
+            Set<TopicPartition> expiredPartitions = new HashSet<>();
+            long currentTimestamp = time.milliseconds();
+            AtomicBoolean hasAllOffsetsExpired = new AtomicBoolean(true);
+            offsetsByTopic.forEach((topic, partitions) -> {
+                if (!expirationCondition.subscribedTopics.contains(topic)) {
+                    partitions.forEach((partition, offsetAndMetadata) -> {
+                        OptionalLong expireTimestampMs = offsetAndMetadata.expireTimestampMs;
+                        if (expireTimestampMs.isPresent()) {
+                            // Older versions with explicit expire_timestamp field => old expiration semantics is used
+                            if (currentTimestamp >= expireTimestampMs.getAsLong()) {
+                                expiredPartitions.add(addOffsetCommitTombstone(groupId, topic, partition, records));
+                            }
+                            // Current version with no per partition retention
+                        } else if (currentTimestamp - expirationCondition.baseTimestamp.apply(offsetAndMetadata) >= offsetsRetentionMs) {
+                            expiredPartitions.add(addOffsetCommitTombstone(groupId, topic, partition, records));
+                        } else {
+                            hasAllOffsetsExpired.set(false);
+                        }
+                    });
+                }
+            });
+            log.debug("[GroupId {}] Expiring offsets: {}", groupId, expiredPartitions);
+
+            if (hasAllOffsetsExpired.get()) {
+                // All offsets were expired for this group. Remove the group.
+                return Optional.of(groupId);
+            }
+
+        } catch (GroupIdNotFoundException e) {
+            // groups in offsets should exist.
+            log.warn("GroupId {} should exist.", groupId);
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Add an offset commit tombstone record for the group.
+     *
+     * @param groupId   The group id.
+     * @param topic     The topic name.
+     * @param partition The partition.
+     * @param records   The list of records to append the tombstone.
+     *
+     * @return The topic partition of the corresponding tombstone.
+     */
+    private TopicPartition addOffsetCommitTombstone(
+        String groupId,
+        String topic,
+        int partition, 
+        List<Record> records
+    ) {
+        records.add(RecordHelpers.newOffsetCommitTombstoneRecord(groupId, topic, partition));
+        TopicPartition tp = new TopicPartition(topic, partition);
+        log.trace("[GroupId {}] Removing expired offset and metadata for {}", groupId, tp);
```

Comment:
nit: We could replace `{}` by `{}-{}` and `tp` by `topic, partition`. This avoid having to allocate a TopicPartition just for the logging.

Reply:
The TopicPartition was created and returned in this method so that we can log all of the expired partitions at the end